File: JS/LC_20.js
/* 
 * Leet code problem 20 - Valid parenthesis
 * Given a string s containing just the characters '(', ')', '{', '}', '[' and ']'
 * determine if the input string is valid. An input string is valid if:
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 * Every close bracket has a corresponding open bracket of the same type.
 * return boolean
 * 
 * Input: s = "{[]}"
 * Output: true
 * Input: s = "(]"
 * Output: false
 */

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    //for odd length return false
    if(s.length%2 !==0) return false
    
    const resultStack = [];
    /* //here we check and add all open brackets into stack
    for (let i = 0; i < s.length; i++) {
        let char = s[i];
        if (char === '(' || char === '{' || char === '['){
            resultStack.push(char); //adding all open one's into stack
        }else{
            returnedChar = resultStack.pop(); //when a closed one is there we pop the recent added open
            if(returnedChar === '(' && char !== ')') return false;
            if(returnedChar === '{' && char !== '}') return false;
            if(returnedChar === '[' && char !== ']') return false;
            //if no open were added when checking for closed or if string had only closed ones then pop() would return undefined
            if(returnedChar === undefined) return false;
        }
    }
    return resultStack.length === 0; */

    for (let index = 0; index < s.length; index++) {
        const char = s[index];
        //for each open bracket add it's respective closed bracket into stack
        /* if(char === '('){
            resultStack.push(')');
        }else if(char === '{'){
            resultStack.push('}');
        }else if(char === '['){
            resultStack.push(']');
        }else{
            if(char !== resultStack.pop()){
                return false;
            }
        } */
        switch (char) {
            case '(':
                resultStack.push(')');
                break;
            case '{':
                resultStack.push('}');
                break;
            case '[':
                resultStack.push(']');
                break;
            default:
                if(char !== resultStack.pop()) return false;
                break;
        }
    }
    return (resultStack.length === 0);
};

console.log(isValid('{[]}()')); //true
console.log(isValid('{[]})(')); //false
console.log(isValid('{(')); //false
console.log(isValid(')}]')); //false

File: Java/LC_20.java
package Java;

import java.util.Stack;

/* 
 * Leet code problem 20 - Valid parenthesis
 * Given a string s containing just the characters '(', ')', '{', '}', '[' and ']'
 * determine if the input string is valid. An input string is valid if:
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 * Every close bracket has a corresponding open bracket of the same type.
 * return boolean
 * 
 * Input: s = "{[]}"
 * Output: true
 * Input: s = "(]"
 * Output: false
 */

public class LC_20 {

    public static boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char bracket = s.charAt(i);
            if (bracket == '[' || bracket == '{' || bracket == '(') {
                // insert all the open brackets into stack
                stack.push(bracket);
            } else {
                // if stack is empty return false as we are seeing an closed bracket but there is no open bracket in stack
                if (stack.isEmpty())
                    return false;
                // see for each closed bracket if it's counter open bracket is not there then return false
                char returnedBracket = stack.pop();
                if (returnedBracket == '(' && s.charAt(i) != ')')
                    return false;
                if (returnedBracket == '{' && s.charAt(i) != '}')
                    return false;
                if (returnedBracket == '[' && s.charAt(i) != ']')
                    return false;
            }
        }

        return stack.empty();
    }

    public static void main(String[] args) {
        String s1 = "{[()]}()", s2 = "{(}[]", s3 = "{[(", s4 = ")]}";
        System.out.println(isValid(s1));
        System.out.println(isValid(s2));
        System.out.println(isValid(s3));
        System.out.println(isValid(s4));
    }

}